src/sagemaker/image.py

```diff
@@ -0,0 +1,446 @@
+
+            try:
+                os.makedirs(os.path.dirname(file_path))
+            except os.error:
+                pass
+
+            obj.download_file(file_path)
+
+    def _generate_compose_file(self, command, additional_volumes=None, additional_env_vars=None):
+        session = self.sagemaker_session.boto_session
+        additional_env_vars = additional_env_vars or []
+        additional_volumes = additional_volumes or {}
+        environment = []
+        optml_dirs = set()
+
+        aws_creds = _aws_credentials(session)
+        if aws_creds is not None:
+            environment.extend(aws_creds)
+
+        environment.extend(additional_env_vars)
+
+        if command == 'train':
+            optml_dirs = {'output', 'input'}
+
+        services = {
+            h: self._create_docker_host(h, environment, optml_dirs,
+                                        command, additional_volumes) for h in self.hosts
+        }
+
+        content = {
+            # Some legacy hosts only support the 2.1 format.
+            'version': '2.1',
+            'services': services
+        }
+
+        filename = os.path.join(self.container_root, DOCKER_COMPOSE_FILENAME)
+        yaml_content = yaml.dump(content, default_flow_style=False)
+        logger.info('docker compose file: \n{}'.format(yaml_content))
+        with open(filename, 'w') as f:
+            f.write(yaml_content)
+
+        return content
+
+    def _compose(self, detached=False):
+        compose_cmd = 'nvidia-docker-compose' if self.instance_type == "local_gpu" else 'docker-compose'
+
+        command = [
+            compose_cmd,
+            '-f',
+            os.path.join(self.container_root, DOCKER_COMPOSE_FILENAME),
+            'up',
+            '--build',
+            '--abort-on-container-exit'
+        ]
+
+        if detached:
+            command.append('-d')
+
+        logger.info('docker command: {}'.format(' '.join(command)))
+        return command
+
+    def _create_docker_host(self, host, environment, optml_subdirs, command, volumes):
+        optml_volumes = self._optml_volumes(host, optml_subdirs)
+        optml_volumes.extend(volumes)
+
+        host_config = {
+            'image': self.image,
+            'stdin_open': True,
+            'tty': True,
+            'volumes': [v.map for v in optml_volumes],
+            'environment': environment,
+            'command': command
+        }
+
+        if command == 'serve':
+            host_config.update({
+                'ports': [
+                    '8080:8080'
+                ]
+            })
+
+        return host_config
+
+    def _optml_volumes(self, host, subdirs, single_model_dir=False):
+        """
+        It takes a folder with the necessary files for training and creates a list of opt volumes that
+        the Container needs to start.
+        If args.single_model_dir is True, all the hosts will point the opt/ml/model subdir to the first container.
```

Comment:
Sure, but this isn't reflective of actual distributed training. This feels like we're trying to compensate for a problem in SageMaker distributed learning. Why are we doing this? Can we not do this?